Handle scaladoc for Scala3

BarkingBad opened the pull request:
<!--- The issue this PR addresses -->
Fixes #16527

### Context

The scaladoc was not working for scala 3. I still need help with creating proper integration tests.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck` // warnings for other subprojects I haven't touched
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

BarkingBad commented:
@ljacomet could you take a look? :D

tgodzik commented on subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaPlugin.java:
```diff
@@ -50,27 +51,34 @@ public void apply(Project project) {
     }
 
-    private static void configureScaladoc(final Project project, final SourceSet main) {
+    private static void configureScaladoc(final Project project, final SourceSet main, final TaskProvider<AbstractScalaCompile> compileScala) {
```
`compileScala` doesn't seem to be used, should we make the scaladoc task depend on it if we use Scala 3 ?

BarkingBad replied:
Leftover, will remove

tgodzik commented on subprojects/scala/src/main/java/org/gradle/api/tasks/ScalaRuntime.java:
```diff
@@ -204,7 +207,8 @@ private DefaultExternalModuleDependency getScalaBridgeDependency(String scalaVer
      *
      * @param scalaVersion version of scala to download the compiler for
-     * @return bridge dependency to download
+     * @param zincVersion version of zinc to download the compiler for as fallback for Scala 2
```
```suggestion
     * @param scalaVersion version of scala to download the compiler for
```

tgodzik reviewed (changes requested):
Thanks for the fix! I left 2 comments, but otherwise looks good.

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/GenerateScaladoc.java:
```diff
@@ -69,9 +71,25 @@ public void execute() {
+        String scaladocFqName = scalaVersion.charAt(0) == '3' ? "dotty.tools.scaladoc.Main" : "scala.tools.nsc.ScalaDoc";
+        String scaladocEntryName = scalaVersion.charAt(0) == '3' ? "run" : "process";
+
+        Class<?> scaladocClass = Thread.currentThread().getContextClassLoader().loadClass(scaladocFqName);
```
You can use `scalaClassLoader` here

BarkingBad replied:
I missed it, thanks for pointing it out

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/GenerateScaladoc.java:
```diff
@@ -69,9 +71,25 @@ public void execute() {
+            props.load(scalaClassLoader.getResourceAsStream("compiler.properties"));
+            return props.getProperty("version.number");
+        } catch (IOException e) {
+            throw new IllegalStateException("Unable to determine scala version");
```
Just to be on the safe side, I would log a warning here but return something that starts with 2 to not cause any issue with older scala versions.

BarkingBad replied:
Ok, will change

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/tasks/scala/internal/GenerateScaladoc.java:
```diff
@@ -69,9 +71,25 @@ public void execute() {
         process.invoke(scaladoc, new Object[]{args.toArray(new String[0])});
     }
+
+    private static String getScalaVersion(ClassLoader scalaClassLoader) {
```
I wonder if we could not inject that information in the worker instead of having to load something here. After all, we had to ask the same question earlier.

BarkingBad replied:
I think this is impossible since the plugin is universal (it's just `scala`) and we determine the version at runtime by the scala library dependency, but maybe you have some idea how to workaround it.

ljacomet replied:
The path would be something like: `ScalaBasePlugin` -> `ScalaDoc` -> `ScaladocParameters` -> `GenerateScaladoc`

makes sense?

ljacomet commented on subprojects/scala/src/integTest/groovy/org/gradle/scala/scaladoc/ScalaDocIntegrationTest.groovy:
```diff
@@ -83,4 +83,28 @@ class ScalaDocIntegrationTest extends AbstractIntegrationSpec implements Directo
         outputContains("maxHeapSize=234M")
     }
+
+    def "scaladoc uses scala3"() {
```
Is there anything specific to Scala 3 scaladoc that we could test for here?

BarkingBad replied:
I'll think about that, but I think the output is very alike to other doc tools. We can assert the existence of certain files I think.

ljacomet reviewed (changes requested):
Thanks a lot for working on this!

A couple questions / comments, see below:

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":921645268,"teamCityBuildId":"46103529","teamCityBuildHeadRef":"41c8b61141028c28f6851d757a61925d377b0754","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46103529) for you.

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaPlugin.java:
```diff
@@ -50,16 +51,15 @@ public void apply(Project project) {
         incrementalAnalysisElements.getOutgoing().artifact(
             compileScalaMapping, configurablePublishArtifact -> configurablePublishArtifact.builtBy(compileScala));
+            
+        configureScaladoc(project, main, compileScala);
```
This is what fails the build - there is no third parameter in `configureScaladoc`

BarkingBad replied:
Yeah, my bad, sorry

BarkingBad commented:
Hi, I implemented the path you proposed for injecting whether it is scala 3 version.

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":922892283,"teamCityBuildId":"46129360","teamCityBuildHeadRef":"25f0eb3a67e0283822ea5d305f1676da4c0d544f","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46129360) for you.

BarkingBad commented:
I forgot to run a sanity check previously so unused imports broke the CI, sorry. Now I checked `./gradlew sanityCheck` and `./gradlew scala:quickTest`, both work fine. Previously Java 16 check passed, but failed for: Quick Feedback, Quick Feedback Linux, and Java 8.

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":923004034,"teamCityBuildId":"46132955","teamCityBuildHeadRef":"d45ed2b67c3cf3e4b16965f980cf822fb966cdb5","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46132955) for you.

BarkingBad commented:
Could you bring some insight into why it fails for Java8 and Quick Feedback?

tgodzik commented:
It looks like `ScalaPluginTest.addsScalaDoc3TasksToTheProject`, `IdeaIntegrationTest.addsScalaFacetAndCompilerLibraries`, `IdeaIntegrationTest.addsScalaSdkAndCompilerLibraries` and `IdeaIntegrationTest.addsScalaSdkAndCompilerLibraries` are failing.

BarkingBad commented:
I fixed `ScalaPluginTest` (I think that there was a problem with creating two test directories, which in fact were the very same directory (they were based on `getClass()`), and on windows there were problems with cleaning caches (Linux was working ok).
I also fixed `IdeaIntegrationTest`, which made me think whether we should rethink how we add scaladoc to gradle, because in scala3 there are a lot of extra transitive dependencies from scaladoc added also to compiler classpath, which can eventually lead to some errors, but changing the whole architecture to keep two separate classpaths can be a tedious job. 
Lastly, I don't know why `CancellationIntegrationTest` failed, I have no idea how my changes affected this test.

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":928984620,"teamCityBuildId":"46308371","teamCityBuildHeadRef":"572f3e069205cfcd7cdc6854251bba8243199887","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46308371) for you.

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java:
```diff
@@ -152,6 +153,17 @@ public void setTitle(@Nullable String title) {
+    @Incubating
+    @Optional
+    @Input
+    public Boolean getIsScala3() {
```
Thanks for looking into my suggestion to inject the Scala version down here.
However adding a public field for this feels wrong.

I had a look at the other changes and what could be done instead is reach out to `ScalaBasePlugin.isScala3` method, which would need to be made public.

This would also remove the need for adding a convention mapping - a concept we are trying to move away from.

ljacomet replied:
Actually, reviewing #18347, let's just add the `ScalaRuntime` as an `@Internal` field instead.
See https://github.com/gradle/gradle/pull/18347/files#diff-bd022006e2ee4c0e9471eff5abec3f44a6e40a84e760796df120df9818c9ca3cR51

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java:
```diff
@@ -198,6 +210,7 @@ protected void generate() {
             parameters.getClasspath().from(getClasspath());
             parameters.getOutputDirectory().set(getDestinationDir());
             parameters.getSources().from(getSource());
+            parameters.getIsScala3().set(getIsScala3());
```
As indicated, reach out to the `ScalaBasePlugin` instead

BarkingBad replied:
Removed

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java:
```diff
@@ -279,6 +288,7 @@ private void configureScaladoc(final Project project, final ScalaRuntime scalaRu
             scalaDoc.getConventionMapping().map("destinationDir", (Callable<File>) () -> project.getExtensions().getByType(JavaPluginExtension.class).getDocsDir().dir("scaladoc").get().getAsFile());
             scalaDoc.getConventionMapping().map("title", (Callable<String>) () -> project.getExtensions().getByType(ReportingExtension.class).getApiDocTitle());
             scalaDoc.getConventionMapping().map("scalaClasspath", (Callable<FileCollection>) () -> scalaRuntime.inferScalaClasspath(scalaDoc.getClasspath()));
+            scalaDoc.getConventionMapping().map("isScala3", (Callable<Boolean>) () -> isScala3(scalaDoc.getScalaClasspath()));
```
As indicated, remove this as the information is obtained in another way

BarkingBad replied:
Removed

BarkingBad commented:
Ok, thanks, I'll try to do it today. One more question, do you know why this one test fails with leaking files (or how to use this tip `org.gradle.integtests.fixtures.executer.GradleExecuter.withFileLeakDetection` for Integration tests). I feel a little bit lost with that. As a workaround, I think moving the scala3 test to another file so a new separate directory will be created can fix the problem.

ljacomet commented:
Let me relaunch a test suite and see

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":929187251,"teamCityBuildId":"46313185","teamCityBuildHeadRef":"2f4ee4d13a8a928642c2d12d6ae23693cc73ade0","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46313185) for you.

BarkingBad commented:
Apparently moving the test to a separate class doesn't fix the problem.

ljacomet commented:
@BarkingBad I'll try to look further into this tomorrow.

Note that I will also do a couple edits as I realized on #18347 that we cannot really have the `ScalaRuntime` on the tasks.
But once that PR is merged, it should be easy to replace it.

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":931569023,"teamCityBuildId":"46374596","teamCityBuildHeadRef":"655cb6c1f441a5c18a020a0fd2868b36d5abc92a","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46374596) for you.

ljacomet commented:
@bot-gradle test this

bot-gradle commented:
<!-- {"replyTargetCommentId":931612628,"teamCityBuildId":"46375979","teamCityBuildHeadRef":"655cb6c1f441a5c18a020a0fd2868b36d5abc92a","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [ReadyForMerge build](https://builds.gradle.org/viewQueued.html?itemId=46375979) for you.

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaPlugin.java:
```diff
@@ -70,7 +71,14 @@ private static void configureScaladoc(final Project project, final SourceSet mai
             });
-            scalaDoc.setSource(main.getExtensions().getByType(ScalaSourceDirectorySet.class));
+
+            boolean isScala3 = ScalaRuntimeHelper.findScalaJar(main.getCompileClasspath().getFiles(), "library_3") != null;
```
This is something that I missed earlier and which is an actual blocker for releasing the support.
It actually causes test failures because in some cases, we end up eagerly resolving the `compileClasspath`, so early that it prevents adding any dependencies ...

This will require more work and so effectively postpone Scaladoc support for Scala 3 to Gradle 7.4

ljacomet replied:
To be clear, fixing this is actually non trivial and may not even be possible with the existing lazy APIs of Gradle. So I will be following up and updating here as soon as I dig deeper.

BarkingBad replied:
Also, the problem is that we want to have the autoconfiguring scala plugin for scala2 and 3, which differs in overall API and makes the whole process harder. I was thinking that maybe we should create separate plugins, like `scala` and `scala3` so there won't be all these classpath guessings.

ljacomet commented:
@bot-gradle test QF

bot-gradle commented:
<!-- {"replyTargetCommentId":932285476,"teamCityBuildId":"46396444","teamCityBuildHeadRef":"48266a46aadfa48c9cf9ff2fd9f6c3c34952edce","teamCityBuildStatus":"PENDING","replyApproveReviewId":null} -->
OK, I've already triggered [QuickFeedback build](https://builds.gradle.org/viewQueued.html?itemId=46396444) for you.

ljacomet commented:
Well looks like I missed one way of fixing this which seems elegant enough and gets rid of the problem

@BarkingBad Any chance you can confirm this works on a real project?

ljacomet commented on subprojects/scala/src/main/java/org/gradle/api/tasks/scala/ScalaDoc.java:
```diff
@@ -197,7 +199,13 @@ protected void generate() {
-            parameters.getSources().from(getSource());
+            boolean isScala3 = ScalaRuntimeHelper.findScalaJar(getScalaClasspath(), "library_3") != null;
+            parameters.getIsScala3().set(isScala3);
+            if (isScala3) {
```
This effectively assumes that a custom `ScalaDoc` task on a custom source set would properly wire both the `classpath` and the `sources` and that we can simply here change what is to be considered "sources"

ljacomet commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=46398120) for you.

bot-gradle commented:
[Pre-tested commit build](https://builds.gradle.org/viewLog.html?buildId=46398120) failed.

ljacomet commented:
@bot-gradle test and merge

bot-gradle commented:
OK, I've already triggered [a build](https://builds.gradle.org/viewQueued.html?itemId=46399127) for you.

BarkingBad commented:
Sorry, I was absent this weekend, but I just have tested it on a small project and it is working fine :D

dubinsky commented:
@ljacomet @BarkingBad is this supposed to be working in 7.3-rc-1?
I am getting `scala.MatchError: val <none> (of class dotty.tools.dotc.core.Symbols$NoSymbol$) while compiling ...`

BarkingBad commented:
Hi, it should be working right now. This error indicates some problems with the compiler itself and can be the result of wrongly gathered classpath for scaladoc generation. Could you share the repo so I could take a look, or try to generate scaladoc from the executable (installed manually or via coursier). Also, what scala version are you using, and is it completing the generation? There used to be problems in older scaladoc versions because it was not stable yet, but all these MatchErrors should not break the generation.

dubinsky commented:
> Hmm, I think I know what can cause the problem. I will experiment tomorrow. Thanks for the repo to test on :)

Thank you!

dubinsky commented:
@ljacomet @BarkingBad issue: https://github.com/gradle/gradle/issues/18732
